packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+     * @param data arbitrary caller-generated data that will be associated with this update.
+     */
+    function mint(address receiver, bytes memory data) public nonReentrant returns (uint256 tokenId) {
+        tokenId = mintNextToken(receiver);
+        emit UpdateToken(tokenId, msg.sender, data);
+    }
+
+    /**
+     * @notice Applies a reward-relevant update to an existing token.
+     * @param tokenId the existing tokenId that the update should be applied to.
+     * @param data arbitrary caller-generated data that will be associated with this update.
+     */
+    function updateToken(uint256 tokenId, bytes memory data) public nonReentrant {
+        emit UpdateToken(tokenId, msg.sender, data);
+    }
+
+    /**
+     * @notice Requests a redemption for any tokenId. This can be called by anyone.
+     * @dev if called by someone who doesn't own the token, they are effectively gifting their bond to the owner.
+     * @param tokenId the tokenId the redemption is for.
+     * @param cumulativeRedemptions the cumulative token addresses and amounts that this tokenId is eligible for
+     * at the current timestap. cumulative redemptions that are too low should be considered to be valid.
+     */
+    function submitRedemption(uint256 tokenId, RedemptionAmount[] memory cumulativeRedemptions) public nonReentrant {
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+        require(redemptions[redemptionId].expiryTime == 0, "Redemption already exists");
+        require(ownerOf(tokenId) != address(0), "tokenId is invalid");
+
+        uint256 time = getCurrentTime();
+
+        uint256 finalFee = _getStore().computeFinalFee(address(bondToken)).rawValue;
+        bondToken.safeTransferFrom(msg.sender, address(this), finalFee + bond);
+
+        redemptions[redemptionId] = Redemption({ finalFee: finalFee, expiryTime: time + liveness });
+
+        emit Submitted(tokenId, redemptionId, cumulativeRedemptions, time);
+    }
+
+    /**
+     * @notice Disputes a redemption request.
+     * @dev this will cancel a request if the final fee changes or something causes the optimistic oracle proposal to
+     * fail.
+     * @param tokenId the tokenId the redemption is for.
+     * @param cumulativeRedemptions the cumulative redemptions that were provided in the original request.
+     */
+    function dispute(uint256 tokenId, RedemptionAmount[] memory cumulativeRedemptions) public nonReentrant {
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+
+        // This automatically checks that redemptions[redemptionId] != 0.
+        // Check that it has not passed liveness liveness.
+        Redemption storage redemption = redemptions[redemptionId];
+        require(getCurrentTime() < redemption.expiryTime, "redemption expired or nonexistent");
+        if (redemption.finalFee != _getStore().computeFinalFee(address(bondToken)).rawValue) {
```

Comment:
how would the final fee have changed?

Reply:
For the same reason as the relayer, we can't do this because the OptimisticOracle uses the current final fee at time of request. Unless we do an OO request at the start of each redemption (and not just on disputes) I think we have to handle the case where the final fee changes. Do you agree with this analysis?

Reply:
I see. well this good thing is this is 
 to happen often. the bad thing is this is expensive!

Reply:
Agreed. I don't think there is much we can do about this, though. We did the exact same thing in the bridgepool, albeit I tried to make the code a little simpler here.